source/sessions/snapshot-sessions.rst

```diff
@@ -0,0 +1,353 @@
+server have been overloaded to take a session parameter. If that session was
+started with ``isSnapshot = true`` then all operations using that
+session will share the same snapshot.
+
+ReadConcern changes
+===================
+
+``shapshot`` added to `ReadConcernLevel enumeration <../read-write-concern/read-write-concern.rst#read-concern>`_.`.
+
+Server Commands
+===============
+
+There are no new server commands related to snapshot reads. Instead,
+snapshot reads are implemented by:
+
+1. Saving the ``atClusterTime`` returned by 4.9+ servers for the first find/aggregate operation in a
+   property ``snapshotClusterTime`` of the ``ClientSession`` object. Drivers MUST save the ``atClusterTime``
+   in the ``ClientSession`` object.
+
+2. Passing that ``snapshotClusterTime`` in the ``atClusterTime`` field of the ``readConcern`` field
+   for subsequent snapshot read operations (for find/aggregate/distinct commands).
+
+Server Command Responses
+========================
+
+To support snapshot reads the server returns the ``atClusterTime`` in
+responses it sends to the driver (for both find/aggregate commands).
+
+.. code:: typescript
+
+    {
+        ok : 1 or 0,
+        ... // the rest of the command reply
+        atClusterTime : <BsonTimestamp>
+    }
+
+The ``atClusterTime`` MUST be stored in the ``ClientSession`` to later be passed as the
+``atClusterTime`` field of the ``readConcern`` with ``snapshot`` level field  in subsequent read operations.
+
+Server Errors
+=============
+1. The server may reply to read commands with a ``SnapshotTooOld`` error if the client's ``atClusterTime`` value is not available in the server's history.
+2. The server will return ``InvalidOptions`` error if both ``atClusterTime`` and ``afterClusterTime`` options are set to true.
+
+Snapshot read commands
+======================
+
+For snapshot reads the driver MUST first obtain ``atClusterTime`` from cursor response of find/aggregate command,
+by specifying ``readConcern`` with ``snapshot`` level field, and store it as ``snapshotClusterTime`` in 
+``ClientSession`` object.
+
+.. code:: typescript
+
+    {
+        find : <string>, // or other read command
+        ... // the rest of the command parameters
+        readConcern :
+        {
+            level : "snapshot"
+        }
+    }
+
+For subsequent reads from same snapshot driver MUST send the ``snapshotClusterTime`` saved in
+the ``ClientSession`` as the value of the ``atClusterTime`` field of the
+``readConcern`` with ``snapshot`` level field:
+
+.. code:: typescript
+
+    {
+        find : <string>, // or other read command
+        ... // the rest of the command parameters
+        readConcern :
+        {
+            level : "snapshot",
+            afterClusterTime : <BsonTimestamp>
+        }
+    }
+
+Lists of commands that support obtaining ``atClusterTime``:
+1. find
+2. aggregate
+
+Lists of commands that support snapshot reads ``atClusterTime``:
+1. find
+2. aggregate
+3. distinct
```

Comment:
Huh, that's strange that distinct supports one but not the other.  Can you point me to where that's documented?

Reply:
Thanks for catching this. I got confused with [SERVER-48460](https://jira.mongodb.org/browse/SERVER-48460).
Fixed and updated the tests accordingly.